File: r152/src/producer152.c
/* Round 152 -- a fast, UNTRUSTED witness producer.
 *
 * It emits an L6-CAPCERT-2 file: for each cell, the claimed value and a walk
 * that attains it.  Nothing it writes is taken on faith -- r152/src/checker152.c
 * and r152/src/checker152.py each re-establish both directions from scratch.
 * It therefore prunes with whatever values it likes, including the claims it is
 * trying to witness; a bad prune can only make it fail to find a witness.
 *
 * build: cc -O2 -o producer152.exe producer152.c
 * run:   ./producer152.exe <claims.txt> <out.txt> [node_cap]
 *        claims.txt: one "<b> <d> <a> <bb> <e> <h> <cap>" per line, in the
 *        order the cells should be certified.
 */
#define CHECKER152_NO_MAIN
#include "checker152.c"

int main(int argc, char **argv) {
    if (argc < 3) { fprintf(stderr, "usage: %s <claims.txt> <out.txt> [node_cap]\n", argv[0]); return 3; }
    NODECAP = argc > 3 ? strtoull(argv[3], NULL, 10) : 0;
    common_setup();

    int arg[8192][6], claim[8192], n = 0;
    FILE *f = fopen(argv[1], "r");
    if (!f) { fprintf(stderr, "cannot open %s\n", argv[1]); return 3; }
    while (fscanf(f, "%d %d %d %d %d %d %d", &arg[n][0], &arg[n][1], &arg[n][2],
                  &arg[n][3], &arg[n][4], &arg[n][5], &claim[n]) == 7) {
        if (++n >= 8192) { fprintf(stderr, "too many claims\n"); return 3; }
    }
    fclose(f);
    for (int i = 0; i < n; ++i) ub_add(arg[i], claim[i]);   /* heuristic only */

    FILE *o = fopen(argv[2], "w");
    if (!o) { fprintf(stderr, "cannot write %s\n", argv[2]); return 3; }
    fputs("L6-CAPCERT-2\n", o);
    fputs("# produced by r152/src/producer152.c -- UNTRUSTED, see the checkers\n", o);
    fputs("# cell <b> <d> <a> <bb> <e> <h> <cap> <nports>, then the ports\n", o);
    fputs("# port index = rank in the lexicographic list of the 720\n", o);
    fputs("# permutations of \"123456\"\n", o);

    Cell c;
    int missing = 0;
    for (int i = 0; i < n; ++i) {
        memcpy(c.arg, arg[i], sizeof c.arg);
        /* One attempt only, at the claimed value.  Walking DOWN would cost a
         * fresh exhaustive search per step and tell us nothing the checker
         * cannot work out for itself, so a miss just leaves the cell without a
         * witness and the checker proves the upper bound alone. */
        int target = claim[i], got = -1;
        static int w[900];
        c.cap = target;
        const char *r = search_cell(&c, target);
        if (!strcmp(r, "FOUND")) {
            got = target;
            for (int j = 0; j < target; ++j) w[j] = TRAIL[j];
        } else if (!strcmp(r, "EXHAUSTED")) {
            fprintf(stderr, "CLAIM NOT ATTAINED %d|%d|%d|%d|%d|%d claim=%d "
                    "(no walk that long exists)\n", arg[i][0], arg[i][1],
                    arg[i][2], arg[i][3], arg[i][4], arg[i][5], claim[i]);
        }
        /* No witness found inside the node cap: still emit the cell with an
         * EMPTY witness.  The checker then proves only the UPPER bound, which
         * is the direction the row census actually consumes. */
        int np = (got == claim[i]) ? got : 0;
        if (np == 0) { ++missing;
            fprintf(stderr, "no witness at the claim for %d|%d|%d|%d|%d|%d "
                    "(claim %d, best found %d) -- upper only\n",
                    arg[i][0], arg[i][1], arg[i][2], arg[i][3], arg[i][4],
                    arg[i][5], claim[i], got); }
        (void)r;
        fprintf(o, "cell %d %d %d %d %d %d %d %d\n", arg[i][0], arg[i][1],
                arg[i][2], arg[i][3], arg[i][4], arg[i][5], claim[i], np);
        for (int j = 0; j < np; ++j) fprintf(o, "%s%d", j ? " " : "", w[j]);
        fputc('\n', o);
        fflush(o);
        fprintf(stderr, "%4d/%4d %d|%d|%d|%d|%d|%d cap=%d\n", i + 1, n,
                arg[i][0], arg[i][1], arg[i][2], arg[i][3], arg[i][4],
                arg[i][5], got);
    }
    fclose(o);
    fprintf(stderr, "cells=%d witnessed=%d upper_only=%d\n", n, n - missing, missing);
    return 0;
}

File: r152/src/producer152.py
#!/usr/bin/env python3
"""Round 152 -- the PRODUCER of capacity certificates.

The producer is UNTRUSTED.  Everything it writes is re-established from scratch
by r152/src/certify152.py:

    * the claimed value   cap(K) = C          is not taken on faith,
    * the witness         a walk with C ports is replayed move by move,
    * the exhaustion      no walk has C+1     is re-searched by the checker.

So the producer may use any heuristic it likes.  It uses the checker's own
search in find-first mode, which is the cheapest source of a witness we have;
if it ever produced a wrong witness or a wrong value the checker would say
DISAGREE rather than silently accept it.

Usage:
    producer152.py --cells r152/certs/pilot_cells_152.json \
                   --out   r152/certs/cap_cert_152.json
"""
from __future__ import annotations
import argparse, json, sys, time
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))
import checker152 as CK


def key(args):
    return "|".join(str(x) for x in args)


def produce(cells, node_cap, claimed=None, prune=None):
    # The producer prunes with whatever values it likes -- including the ones
    # it is trying to witness.  A wrong prune can only make it fail to find a
    # witness; it can never make the checker accept a bad one.
    ck = CK.Checker(prune or {}, node_cap)
    out, order = {}, []
    for args in cells:
        cell = tuple(args)
        k = key(args)
        t0 = time.time()
        # find the largest C for which a walk exists, starting from the claim
        # The producer only has to EXHIBIT a walk; proving that no longer walk
        # exists is the checker's job, so it never runs an exhaustive search.
        # It starts from the value claimed elsewhere and, if that is out of
        # reach, walks down until it finds the best walk it can.  A claim that
        # is too LOW leaves the checker reporting DISAGREE; a claim that is too
        # HIGH leaves the witness short, which the checker rejects as
        # WITNESS_BAD.  Neither can be hidden here.
        C, wit = None, None
        c = claimed.get(k) if claimed else None
        if c is None:
            c = 1
            while ck.search(cell, c) == "FOUND":
                C, wit, c = c, list(ck.trail), c + 1
        else:
            while c >= 1:
                if ck.search(cell, c) == "FOUND":
                    C, wit = c, list(ck.trail)
                    break
                c -= 1
        order.append(k)
        out[k] = dict(args=list(args), cap=C, witness=wit,
                      producer_seconds=round(time.time() - t0, 2))
        print(f"  produced {k}: cap={C} ports={len(wit) if wit else 0} "
              f"{out[k]['producer_seconds']}s", flush=True)
    return order, out


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--cells", required=True)
    ap.add_argument("--out", required=True)
    ap.add_argument("--out-text", default=None)
    ap.add_argument("--claimed", action="append", default=[],
                    help="tables whose cc values are the claims to witness")
    ap.add_argument("--prune-table", action="append", default=[],
                    help="extra production tables used ONLY as search heuristics")
    ap.add_argument("--node-cap", type=int, default=2_000_000_000)
    a = ap.parse_args()
    cells = json.load(open(a.cells))
    src_claimed = a.claimed or ["r147/tables/chain_cells_147.json",
                                "r147/tables/heavy_cells_147.json"]
    claimed, prune = {}, {}
    for src in src_claimed + a.prune_table:
        if not (src and Path(src).exists()):
            continue
        tbl = json.load(open(src))
        for k, v in tbl.items():
            prune[tuple(int(x) for x in k.split("|"))] = v["cc"]
            if src in src_claimed:
                claimed[k] = v["cc"]
    order, cellmap = produce(cells, a.node_cap, claimed, prune)
    doc = dict(format="L6-CAPCERT-1", order=order, cells=cellmap)
    Path(a.out).parent.mkdir(parents=True, exist_ok=True)
    json.dump(doc, open(a.out, "w"), indent=1, sort_keys=True)
    print(f"wrote {a.out}: {len(order)} cells")
    if a.out_text:
        write_text(doc, a.out_text)
        print(f"wrote {a.out_text}")


def write_text(doc, path):
    """The plain-text certificate both checkers read.

    L6-CAPCERT-2
    cell <b> <d> <a> <bb> <e> <h> <cap> <nports>
    <nports port indices>

    A port index is the position of the permutation in the LEXICOGRAPHIC list
    of the 720 permutations of the string "123456"; both checkers rebuild that
    list themselves, so the file carries no table of its own.
    """
    with open(path, "w") as f:
        f.write("L6-CAPCERT-2\n")
        f.write("# cell <b> <d> <a> <bb> <e> <h> <cap> <nports>, then the ports\n")
        f.write("# port index = rank of the permutation in the lexicographic\n")
        f.write("# list of the 720 permutations of \"123456\"\n")
        for k in doc["order"]:
            r = doc["cells"][k]
            w = r["witness"]
            f.write("cell " + " ".join(str(x) for x in r["args"]) +
                    f" {r['cap']} {len(w)}\n")
            f.write(" ".join(str(x) for x in w) + "\n")


if __name__ == "__main__":
    main()
